src/Init/Lean/PrettyPrinter/Parenthesizer.lean

```diff
@@ -24,8 +24,9 @@ Recall that a Pratt parser greedily parses a leading token with precedence at le
 by zero or more trailing tokens with precedence *higher* than `rbp`. Thus we should parenthesize a syntax node `stx`
 produced by `p rbp` if
 
-1. the left-most token in `stx` has precedence < `rbp`, or
-2. the left-most token to *the right of* `stx` has precedence > `rbp` (because without parentheses, `p rbp` would have
+1. the leading/any trailing token in `stx` has precedence < `rbp`/<= `rbp`, respectively (because without parentheses,
+   `p rbp` would not have parsed all of `stx`), or
```

Comment:
This is the one commit where I had to fundamentally revise the parenthesizer algorithm. Surprisingly complex, again!